:farmer: ros2cli_daemon tests failing in win rel because it can't spawn daemons

### Generated by Generative AI

_No response_

### Operating System:

Windows 11

### ROS version or commit hash:

Rolling

### RMW implementation (if applicable):

_No response_

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

1. Go to https://ci.ros2.org/job/nightly_win_rel
2. Build a job
3. See test failing

### Expected behavior

Build without test regressions

### Actual behavior

Building with lots of test regressions because `assert spawn_daemon()` isn't spawning the daemon

### Additional information

### Reference Builds
- [nightly_win_rel#3794](https://ci.ros2.org/job/nightly_win_rel/3794)

### Failing tests
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_name
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_namespace
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_node_names_and_namespaces
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_node_names_and_namespaces_with_enclaves
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_topic_names_and_types
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_service_names_and_types
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_action_names_and_types
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_publisher_names_and_types_by_node
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_subscriber_names_and_types_by_node
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_service_names_and_types_by_node
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_client_names_and_types_by_node
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_action_server_names_and_types_by_node
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_action_client_names_and_types_by_node
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_publishers_info_by_topic
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_subscriptions_info_by_topic
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_clients_info_by_service
ros2cli.ros2cli.test.test_ros2cli_daemon.test_get_servers_info_by_service
ros2cli.ros2cli.test.test_ros2cli_daemon.test_count_publishers
ros2cli.ros2cli.test.test_ros2cli_daemon.test_count_subscribers
ros2cli.ros2cli.test.test_ros2cli_daemon.test_count_clients
ros2cli.ros2cli.test.test_ros2cli_daemon.test_count_services
ros2cli.ros2cli.test.test_strategy.test_with_daemon_running
ros2cli.ros2cli.test.test_strategy.test_with_daemon_spawn
ros2cli.ros2cli.test.test_strategy.test_enforce_no_daemon


### Log Output
```
__________________ ERROR at setup of test_count_subscribers ___________________

test\test_ros2cli_daemon.py:134: in daemon_node

    assert spawn_daemon(args=[], timeout=5.0)

E   assert False

E    +  where False = spawn_daemon(args=[], timeout=5.0)

____________________ ERROR at setup of test_count_clients _____________________

test\test_ros2cli_daemon.py:134: in daemon_node

    assert spawn_daemon(args=[], timeout=5.0)

E   assert False

E    +  where False = spawn_daemon(args=[], timeout=5.0)

____________________ ERROR at setup of test_count_services ____________________

test\test_ros2cli_daemon.py:134: in daemon_node

    assert spawn_daemon(args=[], timeout=5.0)

E   assert False

E    +  where False = spawn_daemon(args=[], timeout=5.0)

_________________ ERROR at setup of test_with_daemon_running __________________

test\test_strategy.py:35: in enforce_daemon_is_running

    assert spawn_daemon(args=[], timeout=5.0)

E   assert False

E    +  where False = spawn_daemon(args=[], timeout=5.0)

__________________ ERROR at setup of test_enforce_no_daemon ___________________

test\test_strategy.py:35: in enforce_daemon_is_running

    assert spawn_daemon(args=[], timeout=5.0)

E   assert False

E    +  where False = spawn_daemon(args=[], timeout=5.0)

================================== FAILURES ===================================

___________________________ test_with_daemon_spawn ____________________________

test\test_strategy.py:54: in test_with_daemon_spawn

    assert node._daemon_node is not None

E   assert None is not None

E    +  where None = <ros2cli.node.strategy.NodeStrategy object at 0x0000023BB9ECB320>._daemon_node
```

### Flakyness 15 days
job_name|last_fail|first_fail|build_count|failure_count|failure_percentage
|--|--|--|--|--|--|
nightly_win_rel|2026-05-11|2026-05-02|4|3|75.0

The first time this error happened was in https://ci.ros2.org/job/nightly_win_rel/3783, comparing repos there is just one change in `ros/ros_environment` and doesn't seems related

CC: @Crola1702

--- a/ros2cli/ros2cli/node/daemon.py
+++ b/ros2cli/ros2cli/node/daemon.py
@@ -75,6 +75,21 @@ def is_daemon_running(args):
         return node.connected
 
 
+def _is_daemon_address_free():
+    # Mirror LocalXMLRPCServer.allow_reuse_address: SO_REUSEADDR on
+    # non-Windows so TIME_WAIT doesn't make us falsely report busy.
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            if os.name != 'nt':
+                s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            s.bind(daemon.get_address())
+        return True
+    except socket.error as e:
+        if e.errno == errno.EADDRINUSE:
+            return False
+        raise
+
+
 def shutdown_daemon(args, timeout=None):
     """
     Shut down daemon node if it's running.
@@ -92,7 +107,11 @@ def shutdown_daemon(args, timeout=None):
             return False
         node.system.shutdown()
         if timeout is not None:
-            predicate = (lambda: not node.connected)
+            # Both conditions matter: a caller spawning a new daemon
+            # right after this returns needs the address to be free.
+            predicate = (
+                lambda: not node.connected and _is_daemon_address_free()
+            )
             if not wait_for(predicate, timeout):
                 raise RuntimeError(
                     'Timed out waiting for '

--- a/ros2cli/test/test_ros2cli_daemon.py
+++ b/ros2cli/test/test_ros2cli_daemon.py
@@ -149,7 +149,8 @@ def daemon_node():
                 f'daemon failed to discover {TEST_NODE_NAMESPACE}/{TEST_NODE_NAME}'
             )
         yield node
-        node.system.shutdown()
+    # Wait for full teardown so the next module's fixture can rebind.
+    shutdown_daemon(args=[], timeout=5.0)
 
 
 def test_get_name(daemon_node):

